drivers/serial/uart_nrf5e.c

```diff
@@ -0,0 +1,873 @@
+ */
+static int uart_nrf5e_init(struct device *dev)
+{
+	volatile struct _uart *uart = UART_STRUCT(dev);
+	struct device *gpio_dev;
+	int err;
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_TX_PIN,
+				  (GPIO_DIR_OUT | GPIO_PUD_PULL_UP));
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_RX_PIN,
+				  (GPIO_DIR_IN));
+
+	uart->PSELTXD = CONFIG_UART_NRF5E_GPIO_TX_PIN;
+	uart->PSELRXD = CONFIG_UART_NRF5E_GPIO_RX_PIN;
+
+#ifdef CONFIG_UART_NRF5E_FLOW_CONTROL
+
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_RTS_PIN,
+				  (GPIO_DIR_OUT | GPIO_PUD_PULL_UP));
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_CTS_PIN,
+				  (GPIO_DIR_IN));
+
+	uart->PSELRTS = CONFIG_UART_NRF5E_GPIO_RTS_PIN;
+	uart->PSELCTS = CONFIG_UART_NRF5E_GPIO_CTS_PIN;
+	uart->CONFIG = (UART_CONFIG_HWFC_Enabled << UART_CONFIG_HWFC_Pos);
+
+#endif /* CONFIG_UART_NRF5E_FLOW_CONTROL */
+
+	DEV_DATA(dev)->baud_rate = CONFIG_UART_NRF5_BAUD_RATE;
+
+	/* Set baud rate */
+	err = baudrate_set(dev, DEV_DATA(dev)->baud_rate,
+		  DEV_CFG(dev)->sys_clk_freq);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	uart->ENABLE = (UARTE_ENABLE_ENABLE_Enabled << UARTE_ENABLE_ENABLE_Pos);
+
+	uart->EVENTS_ENDTX = 0;
+	uart->EVENTS_ENDRX = 0;
+	uart->EVENTS_RXDRDY = 0;
+
+	dev->driver_api = &uart_nrf5e_driver_api;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	DEV_CFG(dev)->irq_config_func(dev);
+	uart_nrf5e_mask(dev, 0);
+	uart_nrf5e_fifo_rx_setup(dev);
+#endif
+
+	return 0;
+}
+
+/**
+ * @brief Poll the device for input.
+ *
+ * @param dev UART device struct
+ * @param c Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+
+static int uart_nrf5e_poll_in(struct device *dev, unsigned char *c)
+{
+
+#ifndef CONFIG_UART_INTERRUPT_DRIVEN
+	volatile struct _uart *uart = UART_STRUCT(dev);
+
+	if (!uart->EVENTS_RXDRDY) {
+		return -1;
+	}
+
+	uart->EVENTS_RXDRDY = 0;
+
+	do {
+		uart->RXD = (u32_t)&c;
+		uart->RXDMAXCNT = 1;
+		uart->TASKS_STARTRX = 1;
+
+		while (!uart->EVENTS_ENDRX) {
+		}
+
+		uart->EVENTS_ENDRX = 0;
+	} while (uart->RXDAMOUNT == 0);
+
+	return 0;
+
+#else
+	if (uart_nrf5e_fifo_read(dev, c, 1)) {
+		return 0;
+	}
+	return -1;
+#endif
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * Checks if the transmitter is empty. If empty, a character is written to
+ * the data register.
+ *
+ * @param dev UART device struct
+ * @param c Character to send
+ *
+ * @return Sent character
+ */
+
+static unsigned char uart_nrf5e_poll_out(struct device *dev,
+					unsigned char c)
+{
+#ifndef CONFIG_UART_INTERRUPT_DRIVEN
+	volatile struct _uart *uart = UART_STRUCT(dev);
+
+	do {
+		uart->TXD = (u32_t)&c;
+		uart->TXDMAXCNT = 1;
+		uart->TASKS_STARTTX = 1;
+
+		while (!uart->EVENTS_ENDTX) {
+		}
+
+		uart->EVENTS_ENDTX = 0;
+	} while (uart->TXDAMOUNT == 0);
```

Comment:
Hi, has this been proven to work in a stable way? I am wondering a bit, since the TXD AMOUNT register is not _specifically_ mentioned in the UARTE "Transmission" Section in the Product Specification. It is not clear to me (while reading the text or looking at Figure 2) when exactly the register value will be updated, i.e. before or after the ENDTX event. 
Also, I was not able to see in the code, where exactly the TXDAMOUNT register is cleared.

In my opinion, the inner while loop is sufficient (ENDTX event will always mark the end of transmission of a single character). What do you think?